File: OS_kernel/py/09_memory_condensation.py
"""
============================================================
KERNEL AGENTIX OS — 09_memory_condensation.py
Module 9: Memory Condensation
Penyimpanan Pola Setimbang / State Persistence
============================================================
Depends on: corekernel.py
"""
import math, json, copy
from corekernel import AgentixKernel


class MemoryCondensation(AgentixKernel):
    def __init__(self, config=None):
        super().__init__(config)
        self._engrams    = {}
        self._snapshots  = {}
        self._max_engrams = (config or {}).get('memory', {}).get('max_engrams', 64)

    def condense(self, namespace, key, vector, weight=1.0):
        self._engrams.setdefault(namespace, [])
        existing = next((e for e in self._engrams[namespace] if e['key'] == key), None)
        if existing:
            existing['vector'] = [v*0.8 + (vector[i] if i < len(vector) else 0)*0.2
                                  for i, v in enumerate(existing['vector'])]
            existing['weight'] = min(1.0, existing['weight'] + weight * 0.1)
            existing['born']   = self.iteration
            return
        self._engrams[namespace].append({
            'key':    key,
            'vector': list(vector),
            'weight': min(1.0, weight),
            'born':   self.iteration,
        })
        if len(self._engrams[namespace]) > self._max_engrams:
            self._engrams[namespace].sort(key=lambda e: e['weight'])
            self._engrams[namespace].pop(0)

    def recall(self, namespace, query_vector, top_k=3):
        engrams = self._engrams.get(namespace, [])
        scored  = [
            {
                'key':        e['key'],
                'weight':     e['weight'],
                'similarity': self._cosine_sim(query_vector, e['vector']) * e['weight'],
                'engram':     e,
            }
            for e in engrams
        ]
        return sorted(scored, key=lambda x: -x['similarity'])[:top_k]

    def _cosine_sim(self, a, b):
        dot   = sum(a[i] * (b[i] if i < len(b) else 0) for i in range(len(a)))
        norm_a = math.sqrt(sum(v**2 for v in a)) + 1e-12
        norm_b = math.sqrt(sum(v**2 for v in b)) + 1e-12
        return dot / (norm_a * norm_b)

    def best_match_valence(self, namespace, query_vector):
        matches = self.recall(namespace, query_vector, 1)
        return max(0.0, matches[0]['similarity']) if matches else 0.0

    def decay_engrams(self, namespace, decay_rate=0.995, min_weight=0.05):
        if namespace not in self._engrams: return
        self._engrams[namespace] = [
            {**e, 'weight': e['weight'] * decay_rate}
            for e in self._engrams[namespace]
            if e['weight'] * decay_rate >= min_weight
        ]

    def decay_all(self, decay_rate=0.995):
        for ns in list(self._engrams): self.decay_engrams(ns, decay_rate)

    def save_snapshot(self, slot_name, state_object):
        self._snapshots[slot_name] = copy.deepcopy(state_object)

    def load_snapshot(self, slot_name):
        return copy.deepcopy(self._snapshots[slot_name]) if slot_name in self._snapshots else None

    def has_snapshot(self, slot_name): return slot_name in self._snapshots
    def list_snapshots(self):          return list(self._snapshots.keys())

    def is_equilibrium(self, namespace, query_vector, threshold=0.85):
        return int(self.best_match_valence(namespace, query_vector) >= threshold)

    def stats(self, namespace):
        engrams = self._engrams.get(namespace, [])
        avg_w   = sum(e['weight'] for e in engrams) / len(engrams) if engrams else 0.0
        return {
            'count':     len(engrams),
            'avg_weight': avg_w,
            'snapshots': len(self._snapshots),
        }

File: OS_kernel/js/09_memory_condensation.js
/**
 * ============================================================
 * KERNEL AGENTIX OS — 09_memory_condensation.js
 * Module 9: Memory Condensation
 * Penyimpanan Pola Setimbang / State Persistence
 * ============================================================
 * Depends on: corekernel.js
 *
 * Memory Condensation: instead of storing raw history,
 *   we condense patterns into weighted "engrams" —
 *   compressed attractors in valence space.
 * Retrieval: by similarity (dot-product / cosine match).
 * Persistence: engrams decay unless reinforced.
 */
import { AgentixKernel, Vector, Semantic } from './corekernel.js';

class MemoryCondensation extends AgentixKernel {
    constructor(config) {
        super(config);
        this._engrams   = {};   // { namespace: [{ key, vector, weight, born }] }
        this._snapshots = {};   // { slotName: stateObject } — hard state saves
        this._maxEngrams = config?.memory?.maxEngrams || 64;
    }

    // ─────────────────────────────────────────
    // ENGRAM — condensed pattern vector
    // vector: flat array of valences (feature vector)
    // weight: salience / reinforcement strength
    // ─────────────────────────────────────────
    condense(namespace, key, vector, weight = 1.0) {
        if (!this._engrams[namespace]) this._engrams[namespace] = [];

        // Check for existing engram with same key → reinforce
        const existing = this._engrams[namespace].find(e => e.key === key);
        if (existing) {
            // Hebbian-style update: blend toward new vector
            existing.vector = existing.vector.map((v, i) =>
                v * 0.8 + (vector[i] || 0) * 0.2
            );
            existing.weight = Math.min(1.0, existing.weight + weight * 0.1);
            existing.born   = this.iteration;
            return;
        }

        this._engrams[namespace].push({
            key,
            vector: [...vector],
            weight: Math.min(1.0, weight),
            born:   this.iteration,
        });

        // Prune weakest if over capacity
        if (this._engrams[namespace].length > this._maxEngrams) {
            this._engrams[namespace].sort((a, b) => a.weight - b.weight);
            this._engrams[namespace].shift();
        }
    }

    // ─────────────────────────────────────────
    // RECALL — retrieve engrams by cosine similarity
    // Returns sorted matches: [{ key, similarity, weight }]
    // ─────────────────────────────────────────
    recall(namespace, queryVector, topK = 3) {
        const engrams = this._engrams[namespace] || [];
        return engrams
            .map(e => ({
                key:        e.key,
                weight:     e.weight,
                similarity: this._cosineSim(queryVector, e.vector) * e.weight,
                engram:     e,
            }))
            .sort((a, b) => b.similarity - a.similarity)
            .slice(0, topK);
    }

    _cosineSim(a, b) {
        const dot  = a.reduce((s, v, i) => s + v * (b[i] || 0), 0);
        const normA = Math.sqrt(a.reduce((s, v) => s + v*v, 0)) + 1e-12;
        const normB = Math.sqrt(b.reduce((s, v) => s + v*v, 0)) + 1e-12;
        return dot / (normA * normB);
    }

    // ─────────────────────────────────────────
    // BEST MATCH VALENCE
    // How well does current state match any engram?
    // ─────────────────────────────────────────
    bestMatchValence(namespace, queryVector) {
        const matches = this.recall(namespace, queryVector, 1);
        return matches.length > 0 ? Math.max(0, matches[0].similarity) : 0;
    }

    // ─────────────────────────────────────────
    // DECAY — weaken engrams over time
    // Engrams not reinforced fade and are pruned
    // ─────────────────────────────────────────
    decayEngrams(namespace, decayRate = 0.995, minWeight = 0.05) {
        const engrams = this._engrams[namespace];
        if (!engrams) return;
        this._engrams[namespace] = engrams
            .map(e => ({ ...e, weight: e.weight * decayRate }))
            .filter(e => e.weight >= minWeight);
    }

    decayAll(decayRate = 0.995) {
        Object.keys(this._engrams).forEach(ns => this.decayEngrams(ns, decayRate));
    }

    // ─────────────────────────────────────────
    // SNAPSHOT — hard save/load of state objects
    // For resuming simulations between sessions
    // ─────────────────────────────────────────
    saveSnapshot(slotName, stateObject) {
        this._snapshots[slotName] = JSON.parse(JSON.stringify(stateObject));
    }

    loadSnapshot(slotName) {
        return this._snapshots[slotName] ? JSON.parse(JSON.stringify(this._snapshots[slotName])) : null;
    }

    hasSnapshot(slotName) { return slotName in this._snapshots; }

    listSnapshots() { return Object.keys(this._snapshots); }

    // ─────────────────────────────────────────
    // EQUILIBRIUM DETECTOR
    // Is the system near a known stable engram?
    // ─────────────────────────────────────────
    isEquilibrium(namespace, queryVector, threshold = 0.85) {
        return +(this.bestMatchValence(namespace, queryVector) >= threshold);
    }

    // ─────────────────────────────────────────
    // MEMORY STATS
    // ─────────────────────────────────────────
    stats(namespace) {
        const engrams = this._engrams[namespace] || [];
        return {
            count:     engrams.length,
            avgWeight: engrams.reduce((s, e) => s + e.weight, 0) / (engrams.length || 1),
            snapshots: this.listSnapshots().length,
        };
    }
}

export { MemoryCondensation };
